File: android/app/src/main/java/app/pospro/terminal/PosPrinterPlugin.java
package app.pospro.terminal;

import android.Manifest;
import android.app.AlertDialog;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.Intent;
import android.os.Build;
import android.util.Base64;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.io.OutputStream;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Bridge printer Bluetooth Classic/SPP untuk POS-PRO.
 *
 * Mengekspos kontrak yang dibaca src/services/androidPrinterBridge.ts sebagai
 * plugin Capacitor "PosPrinter" (window.Capacitor.Plugins.PosPrinter).
 *
 * Socket disimpan statis (bertahan selama proses hidup) dan
 * PrinterForegroundService menjaga proses tidak dibunuh OS saat aplikasi di
 * background — inilah yang membuat printer tetap tersambung saat kasir pindah
 * ke ShopeeFood/Grab, berbeda dengan Web BLE.
 */
@CapacitorPlugin(
    name = "PosPrinter",
    permissions = {
        @Permission(alias = "bluetooth", strings = {
            Manifest.permission.BLUETOOTH_CONNECT,
            Manifest.permission.BLUETOOTH_SCAN
        })
    }
)
public class PosPrinterPlugin extends Plugin {

    private static final UUID SPP_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
    private static BluetoothSocket socket;
    private static OutputStream out;
    private static volatile String deviceAddress;

    private static boolean isLive() {
        return socket != null && socket.isConnected();
    }

    private BluetoothAdapter adapter() {
        return BluetoothAdapter.getDefaultAdapter();
    }

    private boolean needsRuntimeBt() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.S;
    }

    private boolean hasBt() {
        return getPermissionState("bluetooth") == PermissionState.GRANTED;
    }

    // ── Kontrak ─────────────────────────────────────────────────────────────

    @PluginMethod
    public void isSupported(PluginCall call) {
        call.resolve(new JSObject().put("supported", adapter() != null));
    }

    @PluginMethod
    public void isConnected(PluginCall call) {
        call.resolve(new JSObject().put("connected", isLive()));
    }

    @PluginMethod
    public void selectAndConnect(PluginCall call) {
        BluetoothAdapter a = adapter();
        if (a == null) { reject(call, "Perangkat tidak punya Bluetooth"); return; }
        if (!a.isEnabled()) { reject(call, "Nyalakan Bluetooth terlebih dulu"); return; }
        if (needsRuntimeBt() && !hasBt()) {
            requestPermissionForAlias("bluetooth", call, "afterBtPermission");
            return;
        }
        showPickerAndConnect(call);
    }

    @PermissionCallback
    private void afterBtPermission(PluginCall call) {
        if (!hasBt()) { reject(call, "Izin Bluetooth ditolak"); return; }
        showPickerAndConnect(call);
    }

    @PluginMethod
    public void reconnect(PluginCall call) {
        String addr = call.getString("address", deviceAddress);
        if (addr == null) { call.resolve(new JSObject().put("connected", false)); return; }
        if (needsRuntimeBt() && !hasBt()) { call.resolve(new JSObject().put("connected", false)); return; }
        final String fAddr = addr;
        new Thread(() -> {
            BluetoothDevice device = null;
            try { device = adapter().getRemoteDevice(fAddr); } catch (Exception ignored) {}
            boolean ok = device != null && connectTo(device);
            call.resolve(new JSObject().put("connected", ok));
        }).start();
    }

    @PluginMethod
    public void printBase64(PluginCall call) {
        final String data = call.getString("data");
        if (data == null) { call.resolve(err("data kosong")); return; }
        new Thread(() -> {
            OutputStream stream = out;
            if (stream == null || socket == null || !socket.isConnected()) {
                call.resolve(err("printer belum terhubung"));
                return;
            }
            try {
                byte[] bytes = Base64.decode(data, Base64.DEFAULT);
                int i = 0;
                while (i < bytes.length) {
                    int end = Math.min(i + 256, bytes.length);
                    stream.write(bytes, i, end - i);
                    stream.flush();
                    i = end;
                    Thread.sleep(8);
                }
                call.resolve(new JSObject().put("success", true));
            } catch (Exception e) {
                closeQuietly();
                call.resolve(err(e.getMessage() != null ? e.getMessage() : "gagal mengirim ke printer"));
            }
        }).start();
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        closeQuietly();
        stopKeepAlive();
        call.resolve();
    }

    // ── Internal ────────────────────────────────────────────────────────────

    private void showPickerAndConnect(PluginCall call) {
        final List<BluetoothDevice> devices;
        try {
            devices = new ArrayList<>(adapter().getBondedDevices());
        } catch (Exception e) {
            reject(call, "Tidak bisa membaca perangkat paired");
            return;
        }
        if (devices.isEmpty()) {
            reject(call, "Belum ada printer yang dipasangkan. Pair dulu di Pengaturan Bluetooth Android.");
            return;
        }
        getActivity().runOnUiThread(() -> {
            String[] names = new String[devices.size()];
            for (int i = 0; i < devices.size(); i++) names[i] = safeName(devices.get(i));
            new AlertDialog.Builder(getActivity())
                .setTitle("Pilih Printer")
                .setItems(names, (dialog, which) -> {
                    final BluetoothDevice device = devices.get(which);
                    new Thread(() -> {
                        boolean ok = connectTo(device);
                        if (ok) {
                            call.resolve(new JSObject()
                                .put("success", true).put("connected", true)
                                .put("name", safeName(device)).put("address", device.getAddress()));
                        } else {
                            call.resolve(err("Gagal menyambung ke " + safeName(device)));
                        }
                    }).start();
                })
                .setNegativeButton("Batal", (dialog, which) -> reject(call, "Dibatalkan"))
                .setCancelable(true)
                .show();
        });
    }

    /** Buka socket RFCOMM/SPP di thread pemanggil (bukan main thread). */
    private boolean connectTo(BluetoothDevice device) {
        closeQuietly();
        try {
            adapter().cancelDiscovery();
            BluetoothSocket s = device.createRfcommSocketToServiceRecord(SPP_UUID);
            s.connect();
            socket = s;
            out = s.getOutputStream();
            deviceAddress = device.getAddress();
            startKeepAlive();
            return true;
        } catch (Exception first) {
            // Fallback refleksi untuk printer yang rewel dengan UUID standar.
            try {
                Method m = device.getClass().getMethod("createRfcommSocket", int.class);
                BluetoothSocket s = (BluetoothSocket) m.invoke(device, 1);
                s.connect();
                socket = s;
                out = s.getOutputStream();
                deviceAddress = device.getAddress();
                startKeepAlive();
                return true;
            } catch (Exception second) {
                closeQuietly();
                return false;
            }
        }
    }

    private String safeName(BluetoothDevice device) {
        try {
            String n = device.getName();
            return n != null ? n : device.getAddress();
        } catch (SecurityException e) {
            return device.getAddress();
        }
    }

    private void closeQuietly() {
        try { if (out != null) out.close(); } catch (Exception ignored) {}
        try { if (socket != null) socket.close(); } catch (Exception ignored) {}
        out = null;
        socket = null;
    }

    private void startKeepAlive() {
        Intent intent = new Intent(getContext(), PrinterForegroundService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) getContext().startForegroundService(intent);
        else getContext().startService(intent);
    }

    private void stopKeepAlive() {
        getContext().stopService(new Intent(getContext(), PrinterForegroundService.class));
    }

    private JSObject err(String message) {
        return new JSObject().put("success", false).put("error", message);
    }

    private void reject(PluginCall call, String message) {
        call.resolve(new JSObject().put("success", false).put("connected", false).put("error", message));
    }
}

File: capacitor/android/PosPrinterPlugin.kt
package app.pospro.terminal

import android.Manifest
import android.app.AlertDialog
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.Intent
import android.os.Build
import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import java.io.OutputStream
import java.util.UUID

/**
 * Bridge printer Bluetooth Classic/SPP untuk POS-PRO.
 *
 * Mengekspos kontrak yang dibaca src/services/androidPrinterBridge.ts sebagai
 * plugin Capacitor bernama "PosPrinter" (window.Capacitor.Plugins.PosPrinter).
 *
 * Koneksi disimpan di companion object (bertahan selama proses hidup) dan
 * PrinterForegroundService menjaga proses tidak dibunuh saat aplikasi di
 * background — inilah yang membuat printer tetap tersambung saat kasir pindah
 * ke ShopeeFood/Grab, berbeda dengan Web BLE.
 */
@CapacitorPlugin(
  name = "PosPrinter",
  permissions = [
    Permission(
      alias = "bluetooth",
      strings = [
        Manifest.permission.BLUETOOTH_CONNECT,
        Manifest.permission.BLUETOOTH_SCAN,
      ],
    ),
  ],
)
class PosPrinterPlugin : Plugin() {

  companion object {
    private val SPP_UUID: UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
    private var socket: BluetoothSocket? = null
    private var out: OutputStream? = null
    @Volatile private var deviceAddress: String? = null
    @Volatile private var deviceName: String? = null

    fun isLive(): Boolean = socket?.isConnected == true
  }

  private val adapter: BluetoothAdapter? get() = BluetoothAdapter.getDefaultAdapter()

  private fun needsRuntimeBtPermission(): Boolean = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S

  private fun hasBtPermission(): Boolean =
    getPermissionState("bluetooth") == com.getcapacitor.PermissionState.GRANTED

  // ── Kontrak ────────────────────────────────────────────────────────────────

  @PluginMethod
  fun isSupported(call: PluginCall) {
    call.resolve(JSObject().put("supported", adapter != null))
  }

  @PluginMethod
  fun isConnected(call: PluginCall) {
    call.resolve(JSObject().put("connected", isLive()))
  }

  @PluginMethod
  fun selectAndConnect(call: PluginCall) {
    if (adapter == null) return reject(call, "Perangkat tidak punya Bluetooth")
    if (adapter?.isEnabled != true) return reject(call, "Nyalakan Bluetooth terlebih dulu")
    if (needsRuntimeBtPermission() && !hasBtPermission()) {
      requestPermissionForAlias("bluetooth", call, "afterBtPermission")
      return
    }
    showPickerAndConnect(call)
  }

  @PermissionCallback
  fun afterBtPermission(call: PluginCall) {
    if (!hasBtPermission()) return reject(call, "Izin Bluetooth ditolak")
    showPickerAndConnect(call)
  }

  @PluginMethod
  fun reconnect(call: PluginCall) {
    val addr = call.getString("address") ?: deviceAddress
    if (addr == null) return call.resolve(JSObject().put("connected", false))
    if (needsRuntimeBtPermission() && !hasBtPermission()) {
      return call.resolve(JSObject().put("connected", false))
    }
    val device = try { adapter?.getRemoteDevice(addr) } catch (e: Exception) { null }
      ?: return call.resolve(JSObject().put("connected", false))
    Thread {
      val ok = connectTo(device)
      call.resolve(JSObject().put("connected", ok))
    }.start()
  }

  @PluginMethod
  fun printBase64(call: PluginCall) {
    val data = call.getString("data")
      ?: return call.resolve(err("data kosong"))
    Thread {
      val stream = out
      if (stream == null || socket?.isConnected != true) {
        call.resolve(err("printer belum terhubung")); return@Thread
      }
      try {
        val bytes = Base64.decode(data, Base64.DEFAULT)
        var i = 0
        while (i < bytes.size) {
          val end = minOf(i + 256, bytes.size)
          stream.write(bytes, i, end - i)
          stream.flush()
          i = end
          Thread.sleep(8)
        }
        call.resolve(JSObject().put("success", true))
      } catch (e: Exception) {
        // Anggap koneksi putus; tutup agar reconnect berikutnya bersih.
        closeQuietly()
        call.resolve(err(e.message ?: "gagal mengirim ke printer"))
      }
    }.start()
  }

  @PluginMethod
  fun disconnect(call: PluginCall) {
    closeQuietly()
    stopKeepAlive()
    call.resolve()
  }

  // ── Internal ─────────────────────────────────────────────────────────────

  private fun showPickerAndConnect(call: PluginCall) {
    val bonded = try { adapter?.bondedDevices?.toList() } catch (e: Exception) { null }.orEmpty()
    if (bonded.isEmpty()) {
      return reject(call, "Belum ada printer yang dipasangkan. Pair dulu di Pengaturan Bluetooth Android.")
    }
    activity.runOnUiThread {
      val names = bonded.map { safeName(it) }.toTypedArray()
      AlertDialog.Builder(activity)
        .setTitle("Pilih Printer")
        .setItems(names) { _, which ->
          val device = bonded[which]
          Thread {
            val ok = connectTo(device)
            if (ok) {
              call.resolve(
                JSObject()
                  .put("success", true)
                  .put("connected", true)
                  .put("name", safeName(device))
                  .put("address", device.address),
              )
            } else {
              call.resolve(err("Gagal menyambung ke ${safeName(device)}"))
            }
          }.start()
        }
        .setNegativeButton("Batal") { _, _ -> reject(call, "Dibatalkan") }
        .setCancelable(true)
        .show()
    }
  }

  /** Buka socket RFCOMM/SPP di thread pemanggil (bukan main thread). */
  private fun connectTo(device: BluetoothDevice): Boolean {
    closeQuietly()
    try {
      adapter?.cancelDiscovery()
      val s = device.createRfcommSocketToServiceRecord(SPP_UUID)
      s.connect()
      socket = s
      out = s.outputStream
      deviceAddress = device.address
      deviceName = safeName(device)
      startKeepAlive()
      return true
    } catch (first: Exception) {
      // Fallback refleksi untuk printer yang rewel dengan UUID standar.
      try {
        val m = device.javaClass.getMethod("createRfcommSocket", Int::class.javaPrimitiveType)
        val s = m.invoke(device, 1) as BluetoothSocket
        s.connect()
        socket = s
        out = s.outputStream
        deviceAddress = device.address
        deviceName = safeName(device)
        startKeepAlive()
        return true
      } catch (second: Exception) {
        closeQuietly()
        return false
      }
    }
  }

  private fun safeName(device: BluetoothDevice): String =
    try { device.name ?: device.address } catch (e: SecurityException) { device.address }

  private fun closeQuietly() {
    try { out?.close() } catch (_: Exception) {}
    try { socket?.close() } catch (_: Exception) {}
    out = null
    socket = null
  }

  private fun startKeepAlive() {
    val intent = Intent(context, PrinterForegroundService::class.java)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) context.startForegroundService(intent)
    else context.startService(intent)
  }

  private fun stopKeepAlive() {
    context.stopService(Intent(context, PrinterForegroundService::class.java))
  }

  private fun err(message: String): JSObject =
    JSObject().put("success", false).put("error", message)

  private fun reject(call: PluginCall, message: String) {
    call.resolve(JSObject().put("success", false).put("connected", false).put("error", message))
  }
}
